### PR TITLE
Create pkg-config files from CMake

### LINKED ISSUE #10727: cmake: pkgconfig files not installed
In grpc-1.2.3,
```
[    3s] + mkdir -p build
[    3s] + cd build
[    3s] + /usr/bin/cmake /home/abuild/rpmbuild/BUILD/grpc-1.2.3/. '-GUnix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DCMAKE_INSTALL_LIBDIR:PATH=/usr/lib64 -DCMAKE_BUILD_TYPE=RelWithDebInfo '-DCMAKE_C_FLAGS=-fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -DNDEBUG' '-DCMAKE_CXX_FLAGS=-fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -DNDEBUG' '-DCMAKE_Fortran_FLAGS=-fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -DNDEBUG' '-DCMAKE_EXE_LINKER_FLAGS=-Wl,--as-needed -Wl,--no-undefined -Wl,-z,now' '-DCMAKE_MODULE_LINKER_FLAGS=-Wl,--as-needed -Wl,--no-undefined -Wl,-z,now' '-DCMAKE_SHARED_LINKER_FLAGS=-Wl,--as-needed -Wl,--no-undefined -Wl,-z,now' -DLIB_SUFFIX=64 -DCMAKE_SKIP_RPATH:BOOL=ON -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_STATIC_LIBS:BOOL=OFF -DCMAKE_COLOR_MAKEFILE:BOOL=OFF -DCMAKE_INSTALL_DO_STRIP:BOOL=OFF -DCMAKE_MODULES_INSTALL_DIR=/usr/share/cmake/Modules -DgRPC_PROTOBUF_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DgRPC_ZLIB_PROVIDER=package -DgRPC_USE_PROTO_LITE=BOOL:ON
[...]
+ make install
[...]
+ find /tmp/F|grep pc$
/tmp/F/usr/include/grpc
```
However, the plain top-level Makefile *does* have rules for creating .pc files, but the cmake thing does not‽

### PR DESCRIPTION
**Pull request by zackgalbreath**
Fixes #10727 

@nanahpang
@jtattermusch

### CODE REVIEW

**Comment by jtattermusch** [comment]
@zackgalbreath  I started the tests. In the meantime, you'll need to sign the CLA.

File: templates/CMakeLists.txt.template [hunk 1]
```diff
@@ -76,7 +76,7 @@
   set(PACKAGE_STRING    "<%text>${PACKAGE_NAME} ${PACKAGE_VERSION}</%text>")
   set(PACKAGE_TARNAME   "<%text>${PACKAGE_NAME}-${PACKAGE_VERSION}</%text>")
   set(PACKAGE_BUGREPORT "https://github.com/grpc/grpc/issues/")
-  project(<%text>${PACKAGE_NAME}</%text> C CXX)
+  project(<%text>${PACKAGE_NAME}</%text> LANGUAGES C CXX VERSION ${settings.core_version})
```
Comment (jtattermusch) [inline templates/CMakeLists.txt.template]: core_version is  grpc C core version, but the C++ libraries (grpc++) are at a different version (currently 1.24 ish...)

Comment (zackgalbreath) [reply templates/CMakeLists.txt.template]: Yes, I'm using both `grpc_VERSION` (8.0.0) and `PACKAGE_VERSION` (1.24.0-dev) to generate the .pc files from CMake. Is there some reason that you would prefer the CMake project version to be 1.24.0-dev instead?

The [VERSION option of the project() command](https://cmake.org/cmake/help/latest/command/project.html#options) sets some CMake variables that you can use elsewhere in your CMakeLists.txt scripts.

Comment (jtattermusch) [reply templates/CMakeLists.txt.template]: The public project version is the one used for C++ (see our release notes), the C core version number is irrelevant for many users.

That said, can we keep this PR focused on adding the pkgconfig files and perhaps add the project VERSION option in a separate PR - I'm not generally opposed to setting the project VERSION option, but it seems that it's a bit orthogonal to what this PR is trying to achieve (and there are more things to consider when it comes to setting project VERSION).

**Comment by jtattermusch** [comment]
It would also be nice to add some check for the .pc files being setup correctly to one of our distribtests so that this gets continously tested.

File: templates/CMakeLists.txt.template (hunk 1 above)
Comment (zackgalbreath) [reply templates/CMakeLists.txt.template]: 👍

**Comment by jtattermusch** [comment]
@zackgalbreath  have you been able to figure out the CLA signing process?

File: templates/CMakeLists.txt.template [hunk 2]
```diff
@@ -73,10 +73,11 @@
 
   set(PACKAGE_NAME      "grpc")
   set(PACKAGE_VERSION   "${settings.cpp_version}")
+  set(CORE_VERSION      "${settings.core_version}")
   set(PACKAGE_STRING    "<%text>${PACKAGE_NAME} ${PACKAGE_VERSION}</%text>")
   set(PACKAGE_TARNAME   "<%text>${PACKAGE_NAME}-${PACKAGE_VERSION}</%text>")
   set(PACKAGE_BUGREPORT "https://github.com/grpc/grpc/issues/")
-  project(<%text>${PACKAGE_NAME}</%text> C CXX)
+  project(<%text>${PACKAGE_NAME}</%text> LANGUAGES C CXX)
```
Comment (jtattermusch) [inline templates/CMakeLists.txt.template]: nit: is this purely a style improvement?

**Comment by jtattermusch** [comment]
@zackgalbreath  can you please regenerate the CMakeLists.txt to resolve the merge conflict?

It would be good to expand our distribtests to make sure the generated .pc files are actually installed and that they work as expected.
The scripts being run by our distribtests are here: https://github.com/grpc/grpc/tree/master/test/distrib/cpp

File: templates/CMakeLists.txt.template (hunk 2 above)
Comment (zackgalbreath) [reply templates/CMakeLists.txt.template]: At the moment, yes. IMO we should eventually set `(project ... VERSION)` to be either `PACKAGE_VERSION` or `CORE_VERSION`. At that point we will need to use the longer signature of the `project()` command where LANGUAGES are explicitly separated.

File: examples/cpp/helloworld/Makefile [hunk 3]
```diff
@@ -20,14 +20,20 @@ CXX = g++
...
+           -lz\
            -ldl
 else
-LDFLAGS += -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc`\
+LDFLAGS += -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc gpr`\
            -pthread\
            -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed\
+           -laddress_sorting\
+           -lcares\
+           -lz\
```
Comment (zackgalbreath) [inline examples/cpp/helloworld/Makefile]: As you can see, I had to add some libraries to the helloworld Makefile to get it to build. Please let me know if you would've expected this to build cleanly without these modifications.

File: templates/CMakeLists.txt.template (hunk 2 above)
Comment (jtattermusch) [reply templates/CMakeLists.txt.template]: Sg, thanks for the explanation.

File: examples/cpp/helloworld/Makefile (hunk 3 above)
Comment (jtattermusch) [reply examples/cpp/helloworld/Makefile]: I think the reason for the difference it that when built and installed via `Makefile`, address_sorting, cares and zlib are statically linked into libgrpc (as far as I can tell)

With cmake installation (as per the distribtest), we're actually building each of these libraries with `gRPC_*_PROVIDER=package`, so they end up installed as individual libraries (and thus need to be linked).

Comment (jtattermusch) [reply examples/cpp/helloworld/Makefile]: That actually begs the question that if the deps are installed as separate libraries, perhaps the pkgconfig files should know about these dependencies? (which is kind of the point of pkgconfig).

Comment (jtattermusch) [reply examples/cpp/helloworld/Makefile]: Discussed with @nicolasnoble offline - 
the pkg config files need to carry information about what the dependencies are (gpr, address_sorting, cares, zlib) for them to be useful - that should make the original helloworld Makefile to work fine.

Comment (zackgalbreath) [reply examples/cpp/helloworld/Makefile]: 👍 I updated the CMake and Makefile created .pc files to indicate these dependencies. I also updated `Requires.private` to `Requires` so pkg-config will find grpc for you if you ask for grpc++.

**Comment by jtattermusch** [comment]
The EasyCLA check is looking good, glad that it finally worked for you!

**Comment by zackgalbreath** [comment]
Are these four failing checks something I should look into?

File: CMakeLists.txt [hunk 4]
```diff
@@ -25,10 +25,11 @@ cmake_minimum_required(VERSION 3.5.1)
 
 set(PACKAGE_NAME      "grpc")
 set(PACKAGE_VERSION   "1.25.0-dev")
+set(CORE_VERSION      "8.0.0")
```
Comment (jtattermusch) [inline CMakeLists.txt]: nit: should this be gRPC_CORE_VERSION?

**Comment by jtattermusch** [comment]
tasks.distribtest.cpp_linux_x64_jessie_routeguide fails with 
https://source.cloud.google.com/results/invocations/96da7424-2661-4830-be5a-536ef2842063/targets/github%2Fgrpc/tests
```
+ make
protoc -I ../../protos --cpp_out=. ../../protos/helloworld.proto
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o helloworld.pb.o helloworld.pb.cc
protoc -I ../../protos --grpc_out=. --plugin=protoc-gen-grpc=`which grpc_cpp_plugin` ../../protos/helloworld.proto
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o helloworld.grpc.pb.o helloworld.grpc.pb.cc
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_client.o greeter_client.cc
g++ helloworld.pb.o helloworld.grpc.pb.o greeter_client.o -L/usr/local/lib `pkg-config --libs protobuf grpc++` -pthread -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_client
/usr/bin/ld: cannot find -lcares
/usr/bin/ld: cannot find -lz
collect2: error: ld returned 1 exit status
Makefile:44: recipe for target 'greeter_client' failed
```

**Comment by jtattermusch** [comment]
The above error seems to be happening because our Makefile build actually links all the dependent libraries statically (e.g. cares, libz) as part of libgrpc, so e.g. `-lcares` should not be used for the pkgconfig files generated by the Makefile.

File: Makefile [hunk 5]
```diff
@@ -804,16 +804,16 @@ GPR_PC_FILE := $(CORE_PC_TEMPLATE)
 PC_NAME = gRPC
 PC_DESCRIPTION = high performance general RPC framework
 PC_CFLAGS =
-PC_REQUIRES_PRIVATE = gpr $(PC_REQUIRES_GRPC) $(PC_REQUIRES_SECURE)
+PC_REQUIRES = gpr $(PC_REQUIRES_GRPC) $(PC_REQUIRES_SECURE)
 PC_LIBS_PRIVATE = $(PC_LIBS_GRPC) $(PC_LIBS_SECURE)
-PC_LIB = -lgrpc
+PC_LIB = -lgrpc -laddress_sorting -lcares -lz
```
Comment (jtattermusch) [inline Makefile]: for Makefile-based install, these libraries will be linked statically into libgrpc  so the libraries won't exist.

File: CMakeLists.txt (hunk 4 above)
Comment (zackgalbreath) [reply CMakeLists.txt]: 👍

File: Makefile (hunk 5 above)
Comment (zackgalbreath) [reply Makefile]: 👍

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/20519
https://github.com/grpc/grpc/issues/20472

**Comment by nanahpang** [comment]
This change breaks linux/grpc_distribtests_standalone test. https://fusion.corp.google.com/projectanalysis/summary/KOKORO/prod%3Agrpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_distribtests_standalone

**Comment by jtattermusch** [comment]
@nanahpang  I think you're wrong.
According to the build history the distribtests were green for a few more builds after merging this PR but they broke once https://github.com/grpc/grpc/pull/20380/files got merged.

The first failing build is https://source.cloud.google.com/results/invocations/3323a229-9ae7-4c3a-abd3-3ef410cacd9d/targets

This PR got integrated earlier at https://source.cloud.google.com/results/invocations/e821ab54-3f6d-4ac1-859b-5e40a841a63c/targets

**Comment by nanahpang** [comment]
Yes, I have noticed that the failure happened when #20380 got merged. But the changes in #20380 doesn't modify pkg-config. Maybe that PR didn't sync with your changes. I will ask Mark to sync and rebuild that PR first.

**Comment by nicolasnoble** [comment]
LGTM, btw.

**Pull request merged**